src/mpid/ch4/shm/posix/release_gather/release_gather.c

```diff
@@ -81,6 +81,19 @@
         knomial_2 - knomial_2 tree type (ranks are added in order from the right side)
         	    knomial_2 is only supported with non topology aware trees.
 
+    - name        : MPIR_CVAR_REDUCE_INTRANODE_MSG_SIZE_THRESHOLD
+      category    : COLLECTIVE
+      type        : int
+      default     : 2048
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        MPIR_CVAR_REDUCE_INTRANODE_TREE_KVAL and MPIR_CVAR_REDUCE_INTRANODE_TREE_TYPE are used
+        when the message size is smaller than or equal to this threshold;
+        MPIR_CVAR_REDUCE_INTRANODE_TREE_KVAL_LARGE and MPIR_CVAR_REDUCE_INTRANODE_TREE_TYPE_LARGE are
+        used when the message size is larger than this threshold.
```

Comment:
I don't quite like where this is going. The release gather algorithms is growing more complex and costly to maintain. What if we need more branching than just a single message size?

Anyway, please justify why this is necessary. For example, what performance gap we are seeing between, say, k=2 and k=4, for small message and large message respectively. Without those reference, it is difficult to justify whether this is worth the cost of future maintenance.

Reply:
Hi Hui, these are some results I got from osu benchmark for intra-node reduction:
6 ranks, 6 PPN:
![image](https://github.com/user-attachments/assets/55625c84-b11e-4670-978b-ca26ef454c6e)
12 ranks, 12 PPN:
![image](https://github.com/user-attachments/assets/9b7915ce-15b7-4588-b379-a9ab4c6b8488)
24 ranks, 24 PPN:
![image](https://github.com/user-attachments/assets/6813b952-4a67-4854-82e2-4e24183508b3)

From the results, we can observe the performance gap between the best algorithm for small messages (knomial_1, k = 8) and the best algorithm for large messages (kary, k = 2). The performance gap for small messages on 24 PPN is not that obvious but it is significant for 6 and 12 PPN and for large messages.
Also, based on the results, it seems like one threshold is enough. But if we need more branching in the future, this implementation may not be the best solution.

Reply:
Thx